### PR TITLE
Delay mounting on the reportActionView until composer is mounted

### PR DESCRIPTION
**Pull request by parasharrajat**
### Details

I do not consider this the best solution as I think the proper way would be much more complex than this. Let's test it and only merge it if it adds value. https://github.com/Expensify/App/issues/7855#issuecomment-1076917185 

### Fixed Issues

$ https://github.com/Expensify/App/issues/7855

### Tests
1. Open the app on a web browser. 
2. Logout(if logined) and then press `ctrl+shift+R` to hard reload. 
3. Now login with some users. 
4. Switch between many chats. 
5. report messages should not slide from the bottom. 

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
1. Open the app. 
2. Logout(if logined) and then press `ctrl+shift+R` to hard reload (on web) 
3. Now login with a user. 
4. Switch between many chats from LHN.
5. Observe that report messages should not slide from the bottom. 
6. Repeat these on each platform.

- [ ] Verify that no errors appear in the JS console

#### Web

 Video of comparison between staging and dev server. First Staging app can be seen with the issue and secondly the dev server, without issue. Rationally, the staging server is much faster than the dev but we can see that the dev is still not showing that issue.
 

https://user-images.githubusercontent.com/24370807/166943672-38462d45-dd21-4e3a-8533-a98f3b5e4bf5.mp4

 
#### Mobile Web

https://user-images.githubusercontent.com/24370807/166959328-c4fd31ef-88de-4e11-a439-0abb6dda8225.mp4

#### Desktop

https://user-images.githubusercontent.com/24370807/166959795-a625019a-9932-4847-9f4d-822816594772.mp4

#### iOS

https://user-images.githubusercontent.com/24370807/166958714-d13a80b8-c400-46fb-b1f4-038bd5278d29.mp4

#### Android

https://user-images.githubusercontent.com/24370807/166981486-eb86dce5-bb67-455f-be7b-df463c53742d.mp4

### CODE REVIEW

**Comment by marcaaron** [comment]
>I do not consider this the best solution as I think the proper way would be much more complex than this. Let's test it and only merge it if it adds value.

What is the proper solution?

File: src/pages/home/ReportScreen.js [hunk 1]
```diff
@@ -184,7 +189,7 @@ class ReportScreen extends React.Component {
                     style={[styles.flex1, styles.justifyContentEnd, styles.overflowHidden]}
                 >
                     {this.shouldShowLoader() && <FullScreenLoadingIndicator />}
-                    {!this.shouldShowLoader() && (
+                    {!this.shouldShowLoader() && this.state.composerMounted && (
```
Comment (marcaaron) [inline src/pages/home/ReportScreen.js:192]: So we will wait for the text input to mount... but why? Can you please explain this both in the PR's description and probably somewhere as a code comment? It's not clear why we need to do this.

**Review by marcaaron — CHANGES_REQUESTED** [review]

**Comment by parasharrajat** [comment]
> What is the proper solution?

I was thinking that we should remove the remounting on the report screen on each report change and just update the UI when props(reportId) change. But there were a few bugs(don't remember them all) previously that forced us to use your remounting. This will keep the composer mounted all the time.

Like
- Updating the draft message on the composer when the report changes.

File: src/pages/home/ReportScreen.js (hunk 1 above)
Comment (parasharrajat) [reply src/pages/home/ReportScreen.js:192]: I will add the comments when the solution is finalized.

**Comment by marcaaron** [comment]
>we should remove the remounting on the report screen on each report change and just update the UI when props(reportId) change. But there were a few bugs(don't remember them all) previously that forced us to use your remounting

Not sure I agree with that approach either. This means that we would leave several views mounted when they are not in view. If they are not needed then why should we leave them rendered? Improving the performance seems like the best option.

**Comment by marcaaron** [comment]
It is kind of a problem to have several reports mounted at once with the current state of the app. When developers look at the code they have no idea that the report could be "rendered" but not "in view".

**Comment by parasharrajat** [comment]
>  This means that we would leave several views mounted when they are not in view

Hmm. No. Currently, there is on ReportScreen component and when the report changes, we remount it which is expensive. This ReportScreen keeps mounted all the time even if the user is looking at LHN. So my idea was that we prevent the remount. Instead, just update the reportScreen's UI(via rerender) when reportId changes.

**Comment by marcaaron** [comment]
Ok I'm not sure if I'm understanding your original proposal then. Why does remounting the report screen make any difference at all? Simply updating the `ReportActionsView` with new props won't work well without a major redesign as the component is bogged down with a lot of logic in the lifecycle methods.

**Comment by parasharrajat** [comment]
No update for now. I will follow on soon. 

--- 

> Why does remounting the report screen make any difference at all? 

Hmm, remounting and rerendering will have the same effect overall as this is just a design glitch. But rerendering does not show this issue due to the reason that composer height does not change among renders. 

As you said, this can also be done via styles. I will try that out.

**Pull request closed without merging**